File: 0528-random-pick-with-weight/0528-random-pick-with-weight.py
# prefix sum and binary search
# SC- O(n)
# TC- O(n)

class Solution:

    def __init__(self, w: List[int]):
        self.prefix_sums = []
        prefix_sum = 0
        for weight in w:
            prefix_sum += weight
            self.prefix_sums.append(prefix_sum)

    def binary_search(self, target) -> int:
        low = 0
        high = len(self.prefix_sums)-1

        while(low<high):
            mid = low + (high-low)//2
            if(self.prefix_sums[mid]<target):
                low = mid+1
            else:
                high = mid
        return low

    def pickIndex(self) -> int:
        target = random.randint(1, self.prefix_sums[-1])
        return self.binary_search(target)


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()

File: 0528-random-pick-with-weight/0528-random-pick-with-weight.cpp
// SC- O(n)

class Solution {
public:
    vector<int> prefix_sum;
    Solution(vector<int>& w) {
        // TC- O(n)
        // for probability based picking, you can select weights based on prefix sum
        int ps = 0;
        for(int i=0; i<w.size(); i++){
            ps += w[i];
            prefix_sum.push_back(ps);
        }
    }
    
    int pickIndex() {
        // TC- O(logn)
        int i = rand()%prefix_sum.back() +1;
        
        // need the first pefix_sum idx >= i
        int left = 0, right = prefix_sum.size()-1;
        while(left<right){
            int mid = left+(right-left)/2;
            if(prefix_sum[mid]<i)
                left = mid+1;
            else
                right = mid;
        }
        return left;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */